File: epi_judge_python/reverse_bits.py
def reverse_bits(x):
    # Implement this placeholder.
    #lookup = init_look_up()
    return (lookup[x&0xFFFF] << 48)^(lookup[x>>16&0xFFFF] << 32)^(lookup[x>>32&0xFFFF] << 16)^(lookup[x>>48]) 

def init_look_up():
    d = {}
    for x in range(0xFFFF+1):
       rev = x
       for i in range(8):
          rev = swap(rev,0+i,15-i)
       d[x]=rev
    return d


def swap(x,i,j):
    if (x>>i)&1 != (x>>j)&1:
       x ^= 1<<i
       x ^= 1<<j
    return x
lookup = init_look_up()

from test_framework import test_utils_generic_main, test_utils

if __name__ == '__main__':
    test_utils_generic_main.generic_test_main("reverse_bits.tsv", reverse_bits)

File: epi_judge_cpp/reverse_bits.cc
#include <map>
#include <iostream>
using namespace std;

unsigned long swap(unsigned long x, int i, int j) {
  if (((x >> i) & 1) != ((x >> j) & 1)) {
     x ^= 1 << i;
     x ^= 1 << j;
  }
  return x;
}

map <long, long> lookup;

int populateLookup() {
	for ( long x = 0; x < 65536; x++ ) {
		long rev = x;
		for ( int i = 0; i < 8; i++ ) {
			rev = swap(rev, i, 15 - i);
		}
		lookup.insert(pair <long, long> (x, rev));
	}
	return 0;
}

unsigned long ReverseBits(unsigned long x) {
  // Implement this placeholder.
  populateLookup();
  return lookup.find(x >> 48);
}

#include "test_framework/test_utils_generic_main.h"

int main(int argc, char* argv[]) {
  std::vector<std::string> param_names{"x"};
  generic_test_main(argc, argv, param_names, "reverse_bits.tsv", &ReverseBits);
  return 0;
}
